### PR TITLE
coprocesser: refactor encoding and add gbk_chinese_ci map

### LINKED ISSUE #11660: coprocesser: it will report error if use gbk_chinese_ci

### PR DESCRIPTION
**Pull request by xiongjiwei**
Signed-off-by: xiongjiwei <xiongjiwei1996@outlook.com>

### What is changed and how it works?

Close https://github.com/tikv/tikv/issues/11660

- refactor encoding and add gbk_chinese_ci map
- add ASCII and binary encoding

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/tidb_query_datatype/src/codec/collation/encoding/ascii.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+#[derive(Debug)]
+pub struct EncodingAscii;
+
+impl Encoding for EncodingAscii {
+    #[inline]
+    fn decode(data: BytesRef) -> Result<Bytes> {
+        for x in data {
+            if *x >= 0x80_u8 {
```
Comment (zhongzc) [inline components/tidb_query_datatype/src/codec/collation/encoding/ascii.rs:22]:
```suggestion
            if x.is_ascii() {
```

Comment (xiongjiwei) [reply components/tidb_query_datatype/src/codec/collation/encoding/ascii.rs]: updated

File: components/tidb_query_datatype/src/def/field_type.rs [hunk 2]
```diff
@@ -145,20 +146,25 @@ impl fmt::Display for Collation {
...
 
-#[derive(Debug, Clone, Copy)]
+#[derive(PartialEq, Debug, Clone, Copy)]
 pub enum Charset {
     UTF8,
     UTF8Mb4,
     Latin1,
     GBK,
+    Binary,
+    Ascii,
```
Comment (zhongzc) [inline components/tidb_query_datatype/src/def/field_type.rs:156]: Should  `Collation` at L107-116 be updated?

Comment (xiongjiwei) [reply components/tidb_query_datatype/src/def/field_type.rs:156]: nope, this two is the missing charset, just add them here is fine

File: components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs [hunk 3]
```diff
@@ -2,20 +2,37 @@
...
+impl<T: UTF8CompatibleEncoding> Encoding for T {
     #[inline]
     fn decode(data: BytesRef) -> Result<Bytes> {
         match str::from_utf8(data) {
             Ok(v) => Ok(Bytes::from(v)),
-            Err(_) => Err(Error::cannot_convert_string("utf8mb4")),
+            Err(_) => Err(Error::cannot_convert_string(T::NAME)),
         }
     }
+}
```
Comment (andylokandy) [inline components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs:17]: Why add this trait rather than impl `Encoding` on all of them?

Comment (andylokandy) [reply components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs:17]: I think it can't save too much codes.

**Comment by zhongzc** [comment]
I don't know if test cases cover here's changes. Please take a look. @andylokandy

File: components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs (hunk 3 above)
Comment (xiongjiwei) [reply components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs:17]: `EncodingUTF8Mb4`, `EncodingUTF8` and `EncodingLatin1` share the same `decode` implemention, only different at the name. 

Sure it saves little codes, but it is not hard to understand and only have the file scope

**Review by andylokandy — APPROVED** [review]

**Review by zhongzc — APPROVED** [review]

**Pull request merged**